File: examples/ai_agent/lib/ai_agent/agent.ex
defmodule AIAgent.Agent do
  @moduledoc """
  An AI agent backed by a QuickBEAM runtime.

  JS manages conversation state and orchestrates the LLM loop.
  Elixir provides the LLM backend, tool implementations, and streaming delivery.
  """

  @script Path.expand("../../priv/js/agent.ts", __DIR__)

  def child_spec({agent_id, opts}) do
    %{
      id: {__MODULE__, agent_id},
      start: {__MODULE__, :start_link, [agent_id, opts]},
      restart: :transient
    }
  end

  def start_link(agent_id, opts \\ []) do
    llm = Keyword.get(opts, :llm, &default_llm/1)
    llm_stream = Keyword.get(opts, :llm_stream, &default_llm_stream/1)
    llm_with_tools = Keyword.get(opts, :llm_with_tools, &default_llm_with_tools/1)
    tools = Keyword.get(opts, :tools, %{})
    subscriber = Keyword.get(opts, :subscriber, nil)

    QuickBEAM.start(
      name: via(agent_id),
      script: @script,
      handlers: %{
        "llm_complete" => fn [prompt] -> llm.(prompt) end,
        "llm_stream" => fn [prompt] -> llm_stream.(prompt) end,
        "llm_complete_with_tools" => fn [prompt] -> llm_with_tools.(prompt) end,
        "tool_call" => fn [name, args] ->
          case Map.get(tools, name) do
            nil -> %{"error" => "unknown tool: #{name}"}
            func -> func.(args)
          end
        end,
        "status_changed" => fn [status] ->
          if subscriber, do: send(subscriber, {:agent_status, agent_id, status})
          nil
        end,
        "stream_chunk" => fn [chunk] ->
          if subscriber, do: send(subscriber, {:agent_chunk, agent_id, chunk})
          nil
        end,
        "stream_done" => fn [result] ->
          if subscriber, do: send(subscriber, {:agent_done, agent_id, result})
          nil
        end
      }
    )
  end

  def chat(agent_id, message) do
    agent_id |> ensure_started()
    QuickBEAM.call(via(agent_id), "chat", [message])
  end

  def chat_stream(agent_id, message) do
    agent_id |> ensure_started()
    QuickBEAM.call(via(agent_id), "chatStream", [message])
  end

  def chat_with_tools(agent_id, message) do
    agent_id |> ensure_started()
    QuickBEAM.call(via(agent_id), "chatWithTools", [message])
  end

  def configure(agent_id, system_prompt) do
    agent_id |> ensure_started()
    QuickBEAM.call(via(agent_id), "configure", [system_prompt])
  end

  def get_history(agent_id) do
    agent_id |> ensure_started()
    QuickBEAM.call(via(agent_id), "getHistory")
  end

  def get_status(agent_id) do
    agent_id |> ensure_started()
    QuickBEAM.call(via(agent_id), "getStatus")
  end

  def clear_history(agent_id) do
    agent_id |> ensure_started()
    QuickBEAM.call(via(agent_id), "clearHistory")
  end

  defp ensure_started(agent_id) do
    case Registry.lookup(AIAgent.AgentRegistry, agent_id) do
      [{_pid, _}] -> :ok
      [] -> DynamicSupervisor.start_child(AIAgent.AgentSupervisor, {__MODULE__, {agent_id, []}})
    end
  end

  defp via(agent_id) do
    {:via, Registry, {AIAgent.AgentRegistry, agent_id}}
  end

  defp default_llm(%{"system" => system, "messages" => messages}) do
    prompt =
      messages
      |> Enum.map(fn %{"role" => role, "content" => content} -> "#{role}: #{content}" end)
      |> Enum.join("\n")

    "Echo (#{length(messages)} messages, system: #{String.slice(system, 0..30)}...): #{String.slice(prompt, -100..-1//1) || prompt}"
  end

  defp default_llm_stream(%{"system" => _system, "messages" => messages}) do
    user_msgs = Enum.filter(messages, &(&1["role"] == "user"))
    last_msg = (List.last(user_msgs) || %{})["content"] || ""
    response = "Streaming echo: #{last_msg}"
    String.graphemes(response)
  end

  defp default_llm_with_tools(%{"system" => _system, "messages" => messages}) do
    user_msgs = Enum.filter(messages, &(&1["role"] == "user"))
    tool_msgs = Enum.filter(messages, &(&1["role"] == "system" and String.contains?(&1["content"] || "", "Tool results")))
    last_user = (List.last(user_msgs) || %{})["content"] || ""

    cond do
      length(tool_msgs) > 0 ->
        %{"type" => "text", "content" => "Based on tool results: #{last_user}"}

      String.contains?(last_user, "weather") ->
        %{"type" => "tool_calls", "calls" => [%{"name" => "get_weather", "args" => %{"city" => "London"}}]}

      true ->
        %{"type" => "text", "content" => "Echo with tools: #{last_user}"}
    end
  end
end

File: examples/ai_agent/priv/js/agent.ts
interface Message {
  id: string
  role: 'user' | 'assistant' | 'system'
  content: string
  timestamp: number
}

interface ToolCall {
  name: string
  args: Record<string, unknown>
}

interface ToolResult {
  name: string
  result: unknown
}

const state = {
  messages: [] as Message[],
  status: 'idle' as 'idle' | 'thinking' | 'error',
  systemPrompt: 'You are a helpful assistant. Be concise.',
}

function configure(systemPrompt: string) {
  state.systemPrompt = systemPrompt
}

async function chat(userMessage: string): Promise<Message> {
  const userMsg: Message = {
    id: Beam.randomUUIDv7(),
    role: 'user',
    content: userMessage,
    timestamp: Date.now(),
  }
  state.messages.push(userMsg)

  state.status = 'thinking'
  Beam.callSync('status_changed', state.status)

  try {
    const response = await Beam.call('llm_complete', buildPrompt()) as string

    const assistantMsg: Message = {
      id: Beam.randomUUIDv7(),
      role: 'assistant',
      content: response,
      timestamp: Date.now(),
    }
    state.messages.push(assistantMsg)

    state.status = 'idle'
    Beam.callSync('status_changed', state.status)

    return assistantMsg
  } catch (error) {
    state.status = 'error'
    Beam.callSync('status_changed', state.status)
    throw error
  }
}

async function chatStream(userMessage: string): Promise<Message> {
  const userMsg: Message = {
    id: Beam.randomUUIDv7(),
    role: 'user',
    content: userMessage,
    timestamp: Date.now(),
  }
  state.messages.push(userMsg)

  const assistantMsg: Message = {
    id: Beam.randomUUIDv7(),
    role: 'assistant',
    content: '',
    timestamp: Date.now(),
  }
  state.messages.push(assistantMsg)

  state.status = 'thinking'
  Beam.callSync('status_changed', state.status)

  try {
    const chunks = await Beam.call('llm_stream', buildPrompt()) as string[]

    for (const chunk of chunks) {
      assistantMsg.content += chunk
      Beam.callSync('stream_chunk', { id: assistantMsg.id, delta: chunk, content: assistantMsg.content })
    }

    Beam.callSync('stream_done', { id: assistantMsg.id, content: assistantMsg.content })

    state.status = 'idle'
    Beam.callSync('status_changed', state.status)

    return assistantMsg
  } catch (error) {
    state.status = 'error'
    Beam.callSync('status_changed', state.status)
    throw error
  }
}

async function chatWithTools(userMessage: string): Promise<Message> {
  const userMsg: Message = {
    id: Beam.randomUUIDv7(),
    role: 'user',
    content: userMessage,
    timestamp: Date.now(),
  }
  state.messages.push(userMsg)

  state.status = 'thinking'
  Beam.callSync('status_changed', state.status)

  try {
    const maxRounds = 5
    let lastResponse = ''

    for (let round = 0; round < maxRounds; round++) {
      const result = await Beam.call('llm_complete_with_tools', buildPrompt()) as
        | { type: 'text', content: string }
        | { type: 'tool_calls', calls: ToolCall[] }

      if (result.type === 'text') {
        lastResponse = result.content
        break
      }

      const toolResults: ToolResult[] = []
      for (const call of result.calls) {
        const toolResult = await Beam.call('tool_call', call.name, call.args)
        toolResults.push({ name: call.name, result: toolResult })
      }

      state.messages.push({
        id: Beam.randomUUIDv7(),
        role: 'system',
        content: `Tool results: ${JSON.stringify(toolResults)}`,
        timestamp: Date.now(),
      })
    }

    const assistantMsg: Message = {
      id: Beam.randomUUIDv7(),
      role: 'assistant',
      content: lastResponse,
      timestamp: Date.now(),
    }
    state.messages.push(assistantMsg)

    state.status = 'idle'
    Beam.callSync('status_changed', state.status)

    return assistantMsg
  } catch (error) {
    state.status = 'error'
    Beam.callSync('status_changed', state.status)
    throw error
  }
}

function getHistory(): Message[] {
  return state.messages
}

function getStatus() {
  return { status: state.status, messageCount: state.messages.length }
}

function clearHistory() {
  state.messages = []
  state.status = 'idle'
}

function buildPrompt(): { system: string, messages: { role: string, content: string }[] } {
  return {
    system: state.systemPrompt,
    messages: state.messages.map(m => ({ role: m.role, content: m.content })),
  }
}
